[DSCVR-86] Use tracer-provided memfd for language detection

smeso-ddg opened the pull request:
<!--
* Contributors are encouraged to read our [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Please fill the below sections if possible with relevant information or links.
-->
### What does this PR do?

Use metadata provided by tracers via memfd for language detection.

### Motivation

Tracers, when available, are the best way to detect the language used by a process.

### Describe how you validated your changes

unit tests and manual tests against an instrumented program in C++
<!--
Validate your changes before merge, ensuring that:
* Your PR is tested by static / unit / integrations / e2e tests
* Your PR description details which e2e tests cover your changes, if any
* The PR description contains details of how you validated your changes. If you validated changes manually and not through automated tests, add context on why automated tests did not fit your changes validation.

If you want additional validation by a second person, you can ask reviewers to do it. Describe how to set up an environment for manual tests in the PR description. Manual validation is expected to happen on every commit before merge.

Any manual validation step should then map to an automated test. Manual validation should not substitute automation, minus exceptions not supported by test tooling yet.
-->

### Possible Drawbacks / Trade-offs

### Additional Notes
<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

smeso-ddg committed:
pkg/util/kernel: export FindMemFdFilePath

Signed-off-by: Salvatore Mesoraca <salvatore.mesoraca@datadoghq.com>

smeso-ddg committed:
service discovery: add support from tracer metadata

Add functions to check if tracer-generated metadata is available
via memfd and parse the provided data.

Only tracer_language and schema_version fields are parsed as there
is no need for the any other field at the moment.

Signed-off-by: Salvatore Mesoraca <salvatore.mesoraca@datadoghq.com>

smeso-ddg committed:
feat(language_detection): add CPP as a supported language

Signed-off-by: Salvatore Mesoraca <salvatore.mesoraca@datadoghq.com>

smeso-ddg committed:
feat(language_detection): add new privileged detection based on tracers

Signed-off-by: Salvatore Mesoraca <salvatore.mesoraca@datadoghq.com>

smeso-ddg committed:
Add release notes

Signed-off-by: Salvatore Mesoraca <salvatore.mesoraca@datadoghq.com>

brycekahle commented:
Is it expected that `github.com/DataDog/datadog-agent/pkg/discovery/tracermetadata` is imported into the agent binary?

vitkyrka commented on pkg/languagedetection/internal/detectors/testdata/tracer_cpp.data:
Maybe you can add a README describing how these binary files were generated?

vitkyrka commented on pkg/discovery/tracermetadata/tracer_memfd.go:
```diff
@@ -0,0 +1,44 @@
+
+// HasTracerMetadata returns true if tracer metadata
+// is present for a given process
+func HasTracerMetadata(pid int, procRoot string) bool {
```
Is this function used?

vitkyrka replied:
Doesn't seem needed even for DSCVR-83 since it should ideally check that the metadata is actually valid, and for that purpose you can just use `GetTracerMetadata`.

vitkyrka commented on pkg/languagedetection/internal/detectors/tracer.go:
```diff
@@ -0,0 +1,50 @@
+func (i tracerDetector) DetectLanguage(proc model.Process) (model.Language, error) {
+	trMeta, err := tracermetadata.GetTracerMetadata(int(proc.GetPid()), i.hostProc)
+	if err != nil {
+		return model.Language{}, fmt.Errorf("error parsing tracer metadata: %s", err)
```
Can we avoid formatting this error and just return `err` since it's going to happen on pretty much every call to the language detector?

vitkyrka commented on releasenotes/notes/service-discovery-d7381b098aeb19f6.yaml:
```diff
@@ -0,0 +1,11 @@
+---
+features:
+  - |
+    Enable language detection via tracers metadata
```
Does this really warrant an entry in the release notes? It's not mandatory and usually only used for major changes afaik. If you remove it you can add the `changelog/no-changelog` label to suppress CI complaints.

smeso-ddg committed:
drop incorrect msgp argument

Signed-off-by: Salvatore Mesoraca <salvatore.mesoraca@datadoghq.com>

smeso-ddg committed:
Drop HasTracerMetadata

Signed-off-by: Salvatore Mesoraca <salvatore.mesoraca@datadoghq.com>

smeso-ddg committed:
Revert "pkg/util/kernel: export FindMemFdFilePath"

This reverts commit ff92dc930f963474ea50d756886044a529a0c847.

It was only needed by HasTracerMetadata

Signed-off-by: Salvatore Mesoraca <salvatore.mesoraca@datadoghq.com>

vitkyrka commented on pkg/languagedetection/internal/detectors/tracer.go:
```diff
@@ -0,0 +1,50 @@
+		name = model.CPP
+	case "python":
+		name = model.Python
+	case "golang":
```
https://github.com/DataDog/instrumentation-telemetry-api-docs/blob/main/GeneratedDocumentation/ApiDocs/v2/schemas-definition.md#:~:text=One%20of%20the%20following%20values

smeso-ddg committed:
Simplify hot path

Signed-off-by: Salvatore Mesoraca <salvatore.mesoraca@datadoghq.com>

vitkyrka commented on pkg/languagedetection/internal/detectors/tracer.go:
```diff
@@ -0,0 +1,50 @@
+	}
+
+	var name model.LanguageName
+	switch trMeta.TracerLanguage {
```
They are defined in the schema and the tracers should follow the schema when they implement it, so I think we can add the languages.

vitkyrka replied:
Everything that runs with the JVM is reported as java right now from the command line detector, so just reporting jvm as java from here too would be consistent with that.

smeso-ddg committed:
s/golang/go/

Change the string to follow the spec.
The tracer doesn't follow the spec, so we can not use its
output for our test case.

Signed-off-by: Salvatore Mesoraca <salvatore.mesoraca@datadoghq.com>

smeso-ddg committed:
add missing languages

Java has not been added because the tracer can only report "JVM",
not "Java".
It is unclear how to handle the semantic discrepancy.

Signed-off-by: Salvatore Mesoraca <salvatore.mesoraca@datadoghq.com>

smeso-ddg committed:
Add explanatory README alongside testdata

Signed-off-by: Salvatore Mesoraca <salvatore.mesoraca@datadoghq.com>

smeso-ddg committed:
Merge branch 'main' into smeso-ddg/tracermetadata

smeso-ddg committed:
Move tracer-based language detection to privileged package

In this way it won't be included in the core agent.

vitkyrka commented on pkg/languagedetection/internal/detectors/privileged/tracer.go:
```diff
@@ -0,0 +1,58 @@
+		name = model.Node
+	case "ruby":
+		name = model.Ruby
+	default:
```
Add jvm -> model.Java?

vitkyrka commented on pkg/languagedetection/languagemodels/language.go:
```diff
@@ -30,6 +30,9 @@ const (
 	PHP LanguageName = "php"
 
+	// CPP language name.
+	CPP LanguageName = "cpp"
```
Handling for this should also be added to https://github.com/DataDog/datadog-agent/blob/aa6f673b0f0bd9f2dd47127834b03c1ffa578cf2/pkg/collector/corechecks/servicediscovery/language/language_linux.go#L23 (can be done in a separate PR).

vitkyrka commented on pkg/discovery/tracermetadata/tracer_memfd.go:
```diff
@@ -0,0 +1,37 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
```
The start year look like it needs to be changed in all new files.

smeso-ddg committed:
Add Java as language for all JVM detections

smeso-ddg committed:
Fix copyright start year in new files

smeso-ddg committed:
Merge branch 'main' into smeso-ddg/tracermetadata

smeso-ddg committed:
Merge branch 'main' into smeso-ddg/tracermetadata
